File: server/hotrod/src/test/scala/org/infinispan/server/hotrod/HotRodStatsTest.scala
package org.infinispan.server.hotrod

import test.HotRodTestingUtil._
import org.testng.annotations.Test
import org.infinispan.test.fwk.TestCacheManagerFactory
import java.lang.reflect.Method
import org.testng.Assert._
import org.infinispan.manager.EmbeddedCacheManager

/**
 * Tests stats operation against a Hot Rod server.
 * 
 * @author Galder Zamarreño
 * @since 4.1
 */
@Test(groups = Array("functional"), testName = "server.hotrod.HotRodStatsTest")
class HotRodStatsTest extends HotRodSingleNodeTest {

   override def createTestCacheManager: EmbeddedCacheManager = {
      val cfg = hotRodCacheConfiguration()
      cfg.jmxStatistics().enable()
      TestCacheManagerFactory.createClusteredCacheManagerEnforceJmxDomain(jmxDomain, cfg)
   }

   def testStats(m: Method) {
      var bytesRead = 0
      var bytesWritten = 0

      var s = client.stats
//      assertTrue(s.get("timeSinceStart") != "0")
      assertEquals(s.get("currentNumberOfEntries"), "0")
      assertEquals(s.get("totalNumberOfEntries"), "0")
      assertEquals(s.get("stores"), "0")
      assertEquals(s.get("retrievals"), "0")
      assertEquals(s.get("hits"), "0")
      assertEquals(s.get("misses"), "0")
      assertEquals(s.get("removeHits"), "0")
      assertEquals(s.get("removeMisses"), "0")
      bytesRead = assertHigherBytes(bytesRead, s.get("totalBytesRead"))
      // At time of request, no data had been written yet
      assertEquals(s.get("totalBytesWritten"), "0")

      client.assertPut(m)
      s = client.stats
      assertEquals(s.get("currentNumberOfEntries"), "1")
      assertEquals(s.get("totalNumberOfEntries"), "1")
      assertEquals(s.get("stores"), "1")
      bytesRead = assertHigherBytes(bytesRead, s.get("totalBytesRead"))
      bytesWritten = assertHigherBytes(bytesWritten, s.get("totalBytesWritten"))

      assertTrue(s.get("totalBytesRead") != "0")
      assertTrue(s.get("totalBytesWritten") != "0")

      assertSuccess(client.assertGet(m), v(m))
      s = client.stats
      assertEquals(s.get("hits"), "1")
      assertEquals(s.get("misses"), "0")
      assertEquals(s.get("retrievals"), "1")
      bytesRead = assertHigherBytes(bytesRead, s.get("totalBytesRead"))
      bytesWritten = assertHigherBytes(bytesWritten, s.get("totalBytesWritten"))

      client.clear

      s = client.stats
      assertEquals(s.get("totalNumberOfEntries"), "1")
      assertEquals(s.get("currentNumberOfEntries"), "0")
   }

   private def assertHigherBytes(currentBytesRead: Int, bytesStr: String): Int = {
      val bytesRead = bytesStr.toInt
      assertTrue(bytesRead > currentBytesRead)
      bytesRead
   }
}

File: server/hotrod/src/test/java/org/infinispan/server/hotrod/HotRodStatsTest.java
package org.infinispan.server.hotrod;

import static org.infinispan.server.hotrod.test.HotRodTestingUtil.assertSuccess;
import static org.infinispan.server.hotrod.test.HotRodTestingUtil.hotRodCacheConfiguration;
import static org.infinispan.server.hotrod.test.HotRodTestingUtil.v;
import static org.infinispan.test.fwk.TestCacheManagerFactory.configureJmx;
import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertNotEquals;
import static org.junit.jupiter.api.Assertions.assertTrue;

import java.lang.reflect.Method;
import java.util.Map;

import org.infinispan.commons.jmx.MBeanServerLookup;
import org.infinispan.commons.jmx.TestMBeanServerLookup;
import org.infinispan.configuration.cache.ConfigurationBuilder;
import org.infinispan.configuration.global.GlobalConfigurationBuilder;
import org.infinispan.manager.EmbeddedCacheManager;
import org.infinispan.test.fwk.TestCacheManagerFactory;
import org.testng.annotations.Test;

/**
 * Tests stats operation against a Hot Rod server.
 *
 * @author Galder Zamarreño
 * @since 4.1
 */
@Test(groups = "functional", testName = "server.hotrod.HotRodStatsTest")
public class HotRodStatsTest extends HotRodSingleNodeTest {

   private final MBeanServerLookup mBeanServerLookup = TestMBeanServerLookup.create();

   @Override
   public EmbeddedCacheManager createTestCacheManager() {
      ConfigurationBuilder cfg = hotRodCacheConfiguration();
      cfg.statistics().enable();
      GlobalConfigurationBuilder globalCfg = GlobalConfigurationBuilder.defaultClusteredBuilder();
      globalCfg.cacheContainer().statistics(true).metrics().accurateSize(true);
      configureJmx(globalCfg, jmxDomain(), mBeanServerLookup);
      return TestCacheManagerFactory.createClusteredCacheManager(globalCfg, cfg);
   }

   public void testStats(Method m) {
      int bytesRead = 0;
      int bytesWritten = 0;

      Map<String, String> s = client().stats();
      assertEquals("0", s.get("currentNumberOfEntries"));
      assertEquals("0", s.get("stores"));
      assertEquals("0", s.get("retrievals"));
      assertEquals("0", s.get("hits"));
      assertEquals("0", s.get("misses"));
      assertEquals("0", s.get("removeHits"));
      assertEquals("0", s.get("removeMisses"));
      bytesRead = assertHigherBytes(bytesRead, s.get("totalBytesRead"));
      // At time of request, we have only done ping
      bytesWritten = Integer.parseInt(s.get("totalBytesWritten"));
      assertTrue(bytesWritten == 5 || bytesWritten == 6, String.format("Expecting %d in [5,6]", bytesWritten));

      client().assertPut(m);
      s = client().stats();
      assertEquals("1", s.get("currentNumberOfEntries"));
      assertEquals("1", s.get("stores"));
      bytesRead = assertHigherBytes(bytesRead, s.get("totalBytesRead"));
      bytesWritten = assertHigherBytes(bytesWritten, s.get("totalBytesWritten"));

      assertNotEquals("0", s.get("totalBytesRead"));
      assertNotEquals("0", s.get("totalBytesWritten"));

      assertSuccess(client().assertGet(m), v(m));
      s = client().stats();
      assertEquals("1", s.get("hits"));
      assertEquals("0", s.get("misses"));
      assertEquals("1", s.get("retrievals"));
      assertHigherBytes(bytesRead, s.get("totalBytesRead"));
      assertHigherBytes(bytesWritten, s.get("totalBytesWritten"));

      client().clear();

      s = client().stats();
      assertEquals("0", s.get("currentNumberOfEntries"));
   }

   private int assertHigherBytes(int currentBytesRead, String bytesStr) {
      int bytesRead = Integer.parseInt(bytesStr);
      assertTrue(bytesRead > currentBytesRead, String.format("Expecting %d > %d", bytesRead, currentBytesRead));
      return bytesRead;
   }
}
